File: src/com/crossbowffs/quotelock/collections/app/QuoteCollectionFragment.java
package com.crossbowffs.quotelock.collections.app;

import android.content.ContentUris;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Bundle;
import android.view.ContextMenu;
import android.view.LayoutInflater;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.SimpleCursorAdapter;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.ListFragment;
import androidx.loader.app.LoaderManager;
import androidx.loader.content.CursorLoader;
import androidx.loader.content.Loader;

import com.crossbowffs.quotelock.R;
import com.crossbowffs.quotelock.collections.provider.QuoteCollectionContract;
import com.crossbowffs.quotelock.consts.PrefKeys;

/**
 * @author Yubyf
 */
public class QuoteCollectionFragment extends ListFragment implements LoaderManager.LoaderCallbacks<Cursor> {
    private SimpleCursorAdapter mAdapter;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        setHasOptionsMenu(true);
        return inflater.inflate(R.layout.layout_list, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        String[] from = {QuoteCollectionContract.Collections.TEXT, QuoteCollectionContract.Collections.SOURCE};
        int[] to = {R.id.listitem_custom_quote_text, R.id.listitem_custom_quote_source};
        mAdapter = new SimpleCursorAdapter(requireContext(), R.layout.listitem_custom_quote, null, from, to, 0);
        setListAdapter(mAdapter);

        LoaderManager loaderManager = LoaderManager.getInstance(this);
        loaderManager.initLoader(0, null, this);

        ListView listView = getListView();
        registerForContextMenu(listView);
    }

    @Override
    public void onCreateContextMenu(@NonNull ContextMenu menu, @NonNull View v, ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo);
        MenuInflater inflater = requireActivity().getMenuInflater();
        inflater.inflate(R.menu.custom_quote_context, menu);
        menu.findItem(R.id.edit_quote).setVisible(false);
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        AdapterView.AdapterContextMenuInfo info = (AdapterView.AdapterContextMenuInfo) item.getMenuInfo();
        final long rowId = info.id;
        if (item.getItemId() == R.id.delete_quote) {
            deleteQuote(rowId);
            return true;
        }
        return super.onContextItemSelected(item);
    }

    @NonNull
    @Override
    public Loader<Cursor> onCreateLoader(int id, Bundle args) {
        return new CursorLoader(requireContext(),
                QuoteCollectionContract.Collections.CONTENT_URI,
                QuoteCollectionContract.Collections.ALL,
                null, null, null);
    }

    @Override
    public void onLoadFinished(@NonNull Loader<Cursor> loader, Cursor data) {
        mAdapter.changeCursor(data);
    }

    @Override
    public void onLoaderReset(@NonNull Loader<Cursor> loader) {
        mAdapter.changeCursor(null);
    }

    private void deleteQuote(long rowId) {
        Uri uri = ContentUris.withAppendedId(QuoteCollectionContract.Collections.CONTENT_URI, rowId);
        int result = requireContext().getContentResolver().delete(uri, null, null);
        if (result >= 0) {
            requireContext().getSharedPreferences(PrefKeys.PREF_QUOTES, Context.MODE_PRIVATE)
                    .edit()
                    .putBoolean(PrefKeys.PREF_QUOTES_COLLECTION_STATE, false)
                    .apply();
            Toast.makeText(requireContext(), R.string.module_custom_deleted_quote, Toast.LENGTH_SHORT).show();
        }
    }
}

File: app/src/main/java/com/crossbowffs/quotelock/collections/app/QuoteCollectionFragment.kt
package com.crossbowffs.quotelock.collections.app

import android.os.Bundle
import android.view.*
import android.widget.Toast
import androidx.core.os.bundleOf
import androidx.fragment.app.setFragmentResult
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import com.crossbowffs.quotelock.R
import com.crossbowffs.quotelock.collections.database.QuoteCollectionEntity
import com.crossbowffs.quotelock.collections.database.quoteCollectionDatabase
import com.crossbowffs.quotelock.components.BaseQuoteListFragment
import com.crossbowffs.quotelock.components.ContextMenuRecyclerView
import com.crossbowffs.quotelock.components.QuoteListAdapter
import com.crossbowffs.quotelock.consts.PREF_QUOTES_COLLECTION_STATE
import com.crossbowffs.quotelock.data.quotesDataStore
import com.crossbowffs.quotelock.utils.ioScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * @author Yubyf
 */
class QuoteCollectionFragment : BaseQuoteListFragment<QuoteCollectionEntity>() {

    private var collectionsObserver: Job? = null

    @Suppress("UNCHECKED_CAST")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        observeCollections()
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View? {
        setFragmentResult(REQUEST_KEY_COLLECTION_LIST_PAGE,
            bundleOf(BUNDLE_KEY_COLLECTION_SHOW_DETAIL_PAGE to false))
        return super.onCreateView(inflater, container, savedInstanceState)
    }

    override fun onCreateContextMenu(
        menu: ContextMenu,
        v: View,
        menuInfo: ContextMenu.ContextMenuInfo?,
    ) {
        super.onCreateContextMenu(menu, v, menuInfo)
        val inflater = requireActivity().menuInflater
        inflater.inflate(R.menu.custom_quote_context, menu)
        menu.findItem(R.id.edit_quote).isVisible = false
    }

    override fun onContextItemSelected(item: MenuItem): Boolean {
        val info = item.menuInfo as? ContextMenuRecyclerView.ContextMenuInfo
        if (item.itemId == R.id.delete_quote) {
            info?.let {
                ioScope.launch {
                    val result = quoteCollectionDatabase.dao().delete(it.id)
                    if (result >= 0) {
                        quotesDataStore.putBoolean(PREF_QUOTES_COLLECTION_STATE, false)
                        withContext(Dispatchers.Main) {
                            Toast.makeText(requireContext(),
                                R.string.module_custom_deleted_quote,
                                Toast.LENGTH_SHORT)
                                .show()
                        }
                    }
                }
            }
            return true
        }
        return super.onContextItemSelected(item)
    }

    override fun showDetailPage(): Boolean = true

    override fun goToDetailPage() {
        setFragmentResult(REQUEST_KEY_COLLECTION_LIST_PAGE,
            bundleOf(BUNDLE_KEY_COLLECTION_SHOW_DETAIL_PAGE to true))
    }

    @Suppress("UNCHECKED_CAST")
    private fun observeCollections() {
        collectionsObserver?.cancel()
        collectionsObserver = lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                quoteCollectionDatabase.dao().getAll().collect {
                    (recyclerView.adapter as? QuoteListAdapter<QuoteCollectionEntity>)?.submitList(
                        it)
                    scrollToPosition()
                }
            }
        }
    }

    companion object {
        const val REQUEST_KEY_COLLECTION_LIST_PAGE = "collection_list_page"
        const val BUNDLE_KEY_COLLECTION_SHOW_DETAIL_PAGE = "collection_show_detail_page"
    }
}